app/knights.py

```diff
@@ -0,0 +1,30 @@
+def apply_effects(knight: dict) -> None:
+    knight["protection"] = sum(a["protection"] for a in knight["armour"])
+    knight["power"] += knight["weapon"]["power"]
+
+    if knight["potion"] is not None:
+        potion_effects = knight["potion"]["effect"]
+        if "power" in potion_effects:
+            knight["power"] += potion_effects["power"]
+        if "protection" in potion_effects:
+            knight["protection"] += potion_effects["protection"]
+        if "hp" in potion_effects:
+            knight["hp"] += potion_effects["hp"]
+
+
+def knight_battle(knight1: dict, knight2: dict) -> None:
+    knight2["hp"] -= max(0, knight1["power"] - knight2["protection"])
+    knight1["hp"] -= max(0, knight2["power"] - knight1["protection"])
+
+    knight1["hp"] = max(0, knight1["hp"])
+    knight2["hp"] = max(0, knight2["hp"])

```

Comment:
check if it is really needed here

Reply:
well, without it, my tests fail, so I think that it is necessary))
and how can you do it differently?